Documentation for PowerShell exploiter (#1427)
guardicore/monkey: docs/content/reference/exploiters/powershell.md

---
title: "PowerShell"
date: 2021-08-24T12:19:21+03:00
draft: false
tags: ["exploit", "windows"]
---
### Description

PowerShell Remoting exploit brute forces machines via WinRM service using credentials provided by the user

> (see ["configuration"]({{< ref "/usage/configuration" >}}) for instructions) .

Reviewer: Can we add some more information here? I don't love just having one sentence in the description where the user probably expects to get a fairly detailed explanation.

---
title: "PowerShell"
date: 2021-08-24T12:19:21+03:00
draft: false
tags: ["exploit", "windows"]
---

### Description

> PowerShell Remoting is a brute-force exploit that uses PowerShell Remoting Protocol (PSRP) and
> Windows Remote Management(WinRM) services to propagate to a victim.

Reviewer: I'm not sure if this is 100% accurate. PSRP is a protocol and PowerShell Remoting is a method of controlling Windows systems. To say that this is an exploit that uses PSRP may suggest that we're exploiting the protocol, which is false.

Try something like this (this is not the best but all that I could come up with right now):
```suggestion
This exploiter uses brute-force to propagate to a victim through PowerShell Remoting using Windows Remote Management (WinRM).
```

Reviewer 2: I don't think we need to be 100% accurate or precise. We're exploiting remote powershell setups, we don't need to elaborate on which protocols powershell remoting uses, right?

Reviewer: Exactly, no need to mention PSRP at all.

PowerShell Remoting and WinRM listens on the following ports:
1. HTTP: 5895
2. HTTPS: 5896

By default, PowerShell Remoting only allows connections from members of the Administrators group. Sessions are launched
under the user's context, so all operating system access controls applied to individual users and groups continue to apply
to them while connected over PowerShell Remoting.

On private networks, the default Windows Firewall rule for PowerShell Remoting accepts all connections. On public networks,
the default Windows Firewall rule allows PowerShell Remoting connections only from within the same subnet. You have to
explicitly change that rule to open PowerShell Remoting to all connections on a public network.

More on [PowerShell Remoting Protocol]("https://docs.microsoft.com/en-us/powershell/scripting/learn/remoting/winrmsecurity?view=powershell-7.1").

### Windows Remote Management

Windows Remote Management (WinRM) is the Microsoft implementation of WS-Management Protocol, a standard Simple Object
Access Protocol (SOAP)-based, firewall-friendly protocol that allows hardware and operating systems, from different
vendors, to interoperate.

> More on [Windows Remote Management]("https://docs.microsoft.com/en-us/windows/win32/winrm/portal").

Reviewer: This is just details about PSRP and WinRM from the documentation. If someone wants to know about them in detail, they can look up the documentation themselves. We don't need to replicate the information in our documentation.

### Implementation

The exploit brute forces the credentials of PSRP with every possible combination of username and password that
the user provides (see ["configuration"]({{< ref "/usage/configuration" >}})).

#### Credentials list

The PowerShell Remoting Client has ability to use the cached username or/and password from the system we are currently
log in. That means that the exploiter uses the following combination of credentials to propagate to the victim in the order written:

1. Username = None and Password = None; which means that the client we use is going to take the stored credentials
from the system we are using to connect. In order for the user to connect without entering username and password
the victim must have enabled basic authentication, http and no encryption on the victim machine.

2. Username list of usernames and Password = None; brute-force with different usernames and None as
password which means that we use the cached one from the system.

> 3. List of usernames and passwords as plain-text entered by the user.

Reviewer: All of this is just implementation details which, as a user, I don't care about. Mentioning brute-forcing is good enough, IMO.

Reviewer 3: I think this information is good, but the level of detail isn't right. What I mean is, it's good information that the exploiter attempts to use stored credentials and passwords. It's too low level to say "we set Password = None".
